### PR TITLE
Shelan's Adagrams

### PR DESCRIPTION
**Pull request by sheland**
# JS Adagrams
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
--- | ---
What patterns were you able to use from your Ruby knowledge to apply to JavaScript? | I was able to use the same logic from my Ruby version of Adagrams and similar methods such as push, split, delete, use if statements.
Did you need to use different strategies to find information online about JavaScript than you would for Ruby? | Yes, I knew what type of methods I wanted to implement so when I looked online to find information, I knew what I was looking for (such as looking for examples of of how to use the delete method in get the proper syntax)
What was a challenge you were able to overcome on this assignment? | I had a lot of problems with syntax, when and where to use ";" and "," for example. I spent a lot of time addressing syntax challenges but eventually I overcame it by looking at several examples (ex, looking at the structure of if statements to find out if I needed "," or ";" )  
What is your favorite thing about learning a new programming language? | Knowing the basics from a previous language and applying it to a new language has been enjoyable. I've also heard that learning a new language is not as daunting as it seems if you already know a language but I never really understood what that meant. I was surprised to learn that loops, hashes, arrays, if-statements, conditionals etc in JS is somewhat similar to Ruby. I thought I would have to learn from "scratch" but so far, that is not the case. I know things will get more challenging but it is comforting to know Ruby knowledge is applicable to JS. 
What is your least favorite thing about learning a new programming language? | Syntax.
Do you have any recommendations on how we could improve this project for the next cohort? |  So far, so good.

### CODE REVIEW

File: src/adagrams.js [hunk 1]
```diff
@@ -1,8 +1,148 @@
...
   drawLetters() {
-    // Implement this method for wave 1
+    const result = []
+    for (let i = 0; i < 10; i++) {
+      const keys = Object.keys(Adagrams.letters);
+      const curr = keys[Math.floor(Math.random()*keys.length)];
+        result.push(curr);
+        Adagrams.letters[curr] -= 1;
+        if (Adagrams.letters[curr] < 1) {
+          delete Adagrams.letters[curr]
```
Comment (Hamled) [inline src/adagrams.js:39]: This is a reasonable algorithm for picking 10 random letters, however there is a currently a flow with this implementation.

Right now we're using the `Adagrams.letters` object to store the available letters and their counts. However, because `Adagrams` is a variable declared at the "top level" of this file (meaning it is not nested within any functions or classes), it is essentially a global.

And because `Adagrams` is a global variable, then `Adagrams.letters` is as well. As such, the scope and _lifetime_ of this variable are the entire program. The impact of all the preceding, is that each time `Adagrams.drawLetters()` is called the same pool of letters is reused and eventually will run out.

You can see this in action if you play the game (`npm run game`) long enough, I think it takes about 12 rounds before the letters run out.

A simple solution to this would be to make a copy of `Adagrams.letters` at the beginning of this function: `const letters = {...Adagrams.letters};`.

File: src/adagrams.js [hunk 2]
```diff
@@ -1,8 +1,148 @@
...
+      "K": 5,
+      "J": 8, "X": 8,
+      "Q": 10, "Z": 10};
+
+    let words = word.toUpperCase().split('');
+    let scores = words.map(function (char) { //array
+      return scoreHash[char];
+    });
+
+    let totalScore = scores.reduce(function (sum, score) { //add score
```
Comment (Hamled) [inline src/adagrams.js:85]: Love the use of `map` and `reduce` here!

File: src/adagrams.js [hunk 3]
```diff
@@ -1,8 +1,148 @@
...
+
+      for(const word of words) {
+        const wordScore = this.scoreWord(word); //obtain points from scoreWord method
+
+        if (wordScore > bestScore) {
+          bestScore = wordScore;
+          bestWord = word;
+      } else if (wordScore == bestScore && bestWord.length != word.length) { //if same points & different length
+          if (word.length == 10 || (word.length < bestWord.length && bestWord.length != 10)) { //if length is 10 OR length less than bestword's length AND its length is not 10
+          bestScore = wordScore;
```
Comment (Hamled) [inline src/adagrams.js:109]: These two conditional statements are fairly complex on their own. We should pay extra-special attention to indentation in this kind of situation because it can be an aid in comprehending the code.

This is also a good use of comments to give context or just more cleanly write out what a particular line is doing.

**Comment by Hamled** [comment]
# JS Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | :white_check_mark:<br>I'm glad to hear that you're seeing how a lot of what you've learned from Ruby is still applicable to JavaScript (and any programming language)!
Small commits with meaningful commit messages | Mostly. The last commit could probably be broken up into a few smaller pieces.
**Code Requirements** |
`drawLetters` method | :white_check_mark:
Uses appropriate data structure to store the letter distribution | :white_check_mark:
All tests for `drawLetters` pass | Mostly.<br>See my comment in the code for more information on how to make the tests pass (technically the failure is happening in other tests, but the bug is in this function).
`usesAvailableLetters` method | :white_check_mark:
All tests for `usesAvailableLetters` pass | :white_check_mark:
`scoreWord` method | :white_check_mark:
Uses appropriate data structure to store the letter scores | :white_check_mark:
All tests for `scoreWord` pass | :white_check_mark:
`highestScoreFrom` method | :white_check_mark:
Appropriately handles edge cases for tie-breaking logic | :white_check_mark:
All tests for `highestScoreFrom` pass | :white_check_mark:
**Overall** | Generally this is good, although there was a small issue stemming from modifying a global variable in `drawLetters`.
